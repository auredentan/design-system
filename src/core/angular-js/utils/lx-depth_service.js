function LxDepthService() {
    const service = this;

    /////////////////////////////
    //                         //
    //    Private attributes   //
    //                         //
    /////////////////////////////

    /**
     * The initial depth.
     *
     * @type {number}
     */
    // eslint-disable-next-line no-magic-numbers
    let _depth = 1000;

    /////////////////////////////
    //                         //
    //     Public functions    //
    //                         //
    /////////////////////////////

    /**
     * Get the highest depth.
     *
     * @return {number} The highest depth.
     */
    function get() {
        return _depth;
    }

    /**
     * Increase depth.
     */
    function increase() {
        _depth++;
    }

    /////////////////////////////

    service.get = get;
    service.increase = increase;
}

/////////////////////////////

angular.module('lumx.utils.depth').service('LxDepthService', LxDepthService);

/////////////////////////////

export { LxDepthService };
