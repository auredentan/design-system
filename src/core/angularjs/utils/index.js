/* eslint-disable import/no-unassigned-import */

import './depth_service';
import './enter-keypress_directive';
import './event-scheduler_service';
import './focus-on-init_directive';
import './focus-trap_service';
import './utils_service';
