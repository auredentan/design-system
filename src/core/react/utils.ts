import React from 'react';

import get from 'lodash/get';
import has from 'lodash/has';
import isEmpty from 'lodash/isEmpty';
import isString from 'lodash/isString';

/////////////////////////////

type Component = string | React.FC<any> | React.PureComponent<any, any> | React.Component<any, any>;

/////////////////////////////

/**
 * Check if a ReactElement is of the given type.
 *
 * @param  {React.ReactElement} el   The ReactElement we want to check the type of.
 * @param  {string|Component}   type The type we want to check if the ReactElement is of.
 * @return {boolean}            If the ReactElement is of the given type or not.
 */
function isElementOfType(el: React.ReactElement, type: Component): boolean {
    if (!isString(type) && has(type, 'displayName')) {
        type = get(type, 'displayName', get(type, '_reactInternalFiber.elementType.name', 'Component'));
    }

    if (isEmpty(type) && !isString(type)) {
        throw new Error(
            `The type you want to check is not valid. Waiting a component or a string, got ${type.toString()}`,
        );
    }

    if (type === 'text') {
        return isString(el) || el.type === 'span';
    }

    return get(el.type, 'name') === type;
}

/**
 * Check if a ReactElement is a text (i.e. either a pure text node or a <span>).
 * Simply check if the ReactElement is a string or if its type is 'span'.
 *
 * @param  {React.ReactElement} el The ReactElement to check if it's a text.
 * @return {boolean}            If the ReactElement is a text or not.
 */
function isElementText(el: React.ReactElement): boolean {
    return isElementOfType(el, 'text');
}

/////////////////////////////

export { Component, isElementOfType, isElementText };
