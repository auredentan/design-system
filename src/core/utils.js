// eslint-disable-next-line import/no-extraneous-dependencies
import classNames from 'classnames';

import isBoolean from 'lodash/isBoolean';
import isEmpty from 'lodash/isEmpty';
import kebabCase from 'lodash/kebabCase';
import { noop } from 'lodash/noop';

import { CSS_PREFIX, ESCAPE_KEY_CODE } from './constants';

/////////////////////////////
//                         //
//    Private functions    //
//                         //
/////////////////////////////

/**
 * Add a css rule in a given sheet.
 *
 * @param {Element} sheet    The sheet to insert the new rules in.
 * @param {string}  selector The css rules selector.
 * @param {string}  rules    The css rules.
 * @param {number}  index    The css rule index.
 */
function _addCSSRule(sheet, selector, rules, index) {
    if ('insertRule' in sheet) {
        sheet.insertRule(`${selector}{${rules}}`, index);
    } else if ('addRule' in sheet) {
        sheet.addRule(selector, rules, index);
    }
}

/**
 * Get button css rules impacted by primary and secondary colors.
 *
 * @param  {Object} colorPalette The custom color palette.
 * @param  {string} color        Whether to return primary or secondary variants.
 * @return {Array}  The button css rules.
 */
function _getButtonCSSRules(colorPalette, color) {
    return [
        // Default state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-high.${CSS_PREFIX}-button--theme-light
            `,
            rule: `background-color: ${colorPalette[color].N}`,
        },
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-medium
            `,
            rule: `background-color: ${colorPalette[color].L5}`,
        },
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-high.${CSS_PREFIX}-button--theme-dark,
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-medium,
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-low
            `,
            rule: `color: ${colorPalette[color].N}`,
        },
        // Hover state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-high.${CSS_PREFIX}-button--theme-light:hover
            `,
            rule: `background-color: ${colorPalette[color].D1}`,
        },
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-medium:hover,
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-low:hover
            `,
            rule: `background-color: ${colorPalette[color].L4}`,
        },
        // Active state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-high.${CSS_PREFIX}-button--theme-light:active
            `,
            rule: `background-color: ${colorPalette[color].D2}`,
        },
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-medium:active,
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-low:active
            `,
            rule: `background-color: ${colorPalette[color].L3}`,
        },
        // Focus state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-high.${CSS_PREFIX}-button--theme-light[data-focus-visible-added],
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-medium[data-focus-visible-added],
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--color-${color}.${CSS_PREFIX}-button--emphasis-low[data-focus-visible-added]
            `,
            rule: `box-shadow: 0 0 0 2px ${colorPalette[color].L3}`,
        },
    ];
}

/**
 * Get selected button css rules impacted by primary color.
 *
 * @param  {Object} colorPalette The custom color palette.
 * @return {Array}  The selected button css rules.
 */
function _getButtonSelectedCSSRules(colorPalette) {
    return [
        // Default state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--is-selected.${CSS_PREFIX}-button--color-dark
            `,
            rule: `background-color: ${colorPalette.primary.L4}; color: ${colorPalette.primary.D2};`,
        },
        // Hover state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--is-selected.${CSS_PREFIX}-button--color-dark:hover
            `,
            rule: `background-color: ${colorPalette.primary.L3}`,
        },
        // Active state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--is-selected.${CSS_PREFIX}-button--color-dark:active
            `,
            rule: `background-color: ${colorPalette.primary.L2}`,
        },
        // Focus state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-button--is-selected.${CSS_PREFIX}-button--color-dark[data-focus-visible-added]
            `,
            rule: `box-shadow: 0 0 0 2px ${colorPalette.primary.L3}`,
        },
    ];
}

/**
 * Get checkbox css rules impacted by primary or secondary colors.
 *
 * @param  {Object} colorPalette The custom color palette.
 * @param  {string} color        Whether to return primary or secondary variants.
 * @return {Array}  The checkbox css rules.
 */
function _getCheckboxCSSRules(colorPalette, color) {
    return [
        // Default state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-checkbox--theme-light .${CSS_PREFIX}-checkbox__input-indicator
            `,
            rule: `background-color: ${colorPalette[color].N}`,
        },
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-checkbox--theme-dark .${CSS_PREFIX}-checkbox__input-indicator
            `,
            rule: `color: ${colorPalette[color].N}`,
        },
        // Hover state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-checkbox--theme-light.${CSS_PREFIX}-checkbox--is-checked
                .${CSS_PREFIX}-checkbox__input-native:hover
                + .${CSS_PREFIX}-checkbox__input-placeholder .${CSS_PREFIX}-checkbox__input-indicator
            `,
            rule: `background-color: ${colorPalette[color].D1}`,
        },
        // Active state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-checkbox--theme-light.${CSS_PREFIX}-checkbox--is-checked
                .${CSS_PREFIX}-checkbox__input-native:active
                + .${CSS_PREFIX}-checkbox__input-placeholder .${CSS_PREFIX}-checkbox__input-indicator
            `,
            rule: `background-color: ${colorPalette[color].D2}`,
        },
        // Focus state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-checkbox--theme-light.${CSS_PREFIX}-checkbox--is-checked
                .${CSS_PREFIX}-checkbox__input-native[data-focus-visible-added]
                + .${CSS_PREFIX}-checkbox__input-placeholder
            `,
            rule: `box-shadow: 0 0 0 2px ${colorPalette[color].L3}`,
        },
    ];
}

/**
 * Get selected chip css rules impacted by primary color.
 *
 * @param  {Object} colorPalette The custom color palette.
 * @return {Array}  The selected chip css rules.
 */
function _getChipSelectedCSSRules(colorPalette) {
    return [
        // Default state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-chip--is-selected.${CSS_PREFIX}-chip--color-dark
            `,
            rule: `background-color: ${colorPalette.primary.L4}; color: ${colorPalette.primary.D2};`,
        },
        // Hover state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-chip--is-selected.${CSS_PREFIX}-chip--color-dark:hover
            `,
            rule: `background-color: ${colorPalette.primary.L3}`,
        },
        // Active state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-chip--is-selected.${CSS_PREFIX}-chip--color-dark:active
            `,
            rule: `background-color: ${colorPalette.primary.L2}`,
        },
        // Focus state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-chip--is-selected.${CSS_PREFIX}-chip--color-dark[data-focus-visible-added]
            `,
            rule: `box-shadow: 0 0 0 2px ${colorPalette.primary.L3}`,
        },
    ];
}

/**
 * Get progress css rules impacted by primary color.
 *
 * @param  {Object} colorPalette The custom color palette.
 * @return {Array}  The progress css rules.
 */
function _getProgressCSSRules(colorPalette) {
    return [
        {
            selector: `
                .${CSS_PREFIX}-custom-colors .${CSS_PREFIX}-progress-circular__double-bounce1,
                .${CSS_PREFIX}-custom-colors .${CSS_PREFIX}-progress-circular__double-bounce2,
                .${CSS_PREFIX}-custom-colors .${CSS_PREFIX}-progress-linear__line1,
                .${CSS_PREFIX}-custom-colors .${CSS_PREFIX}-progress-linear__line2
            `,
            rule: `background-color: ${colorPalette.primary.L3}`,
        },
    ];
}

/**
 * Get radio button css rules impacted by primary or secondary colors.
 *
 * @param  {Object} colorPalette The custom color palette.
 * @param  {string} color        Whether to return primary or secondary variants.
 * @return {Array}  The radio button css rules.
 */
function _getRadioButtonCSSRules(colorPalette, color) {
    return [
        // Default state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-radio-button--theme-light.${CSS_PREFIX}-radio-button--is-checked .${CSS_PREFIX}-radio-button__input-background
            `,
            rule: `color: ${colorPalette[color].N}`,
        },
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-radio-button--theme-light.${CSS_PREFIX}-radio-button--is-unchecked .${CSS_PREFIX}-radio-button__input-indicator,
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-radio-button--theme-light.${CSS_PREFIX}-radio-button--is-checked .${CSS_PREFIX}-radio-button__input-indicator
            `,
            rule: `background-color: ${colorPalette[color].N}`,
        },
        // Focus state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-radio-button--theme-light.${CSS_PREFIX}-radio-button--is-checked
                .${CSS_PREFIX}-radio-button__input-native[data-focus-visible-added]
                + .${CSS_PREFIX}-radio-button__input-placeholder
            `,
            rule: `box-shadow: 0 0 0 2px ${colorPalette[color].L3}`,
        },
    ];
}

/**
 * Get select css rules impacted by primary color.
 *
 * @param  {Object} colorPalette The custom color palette.
 * @return {Array}  The select css rules.
 */
function _getSelectCSSRules(colorPalette) {
    return [
        // Focus state.
        {
            selector: `
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-select--theme-light.${CSS_PREFIX}-select--is-open .${CSS_PREFIX}-select__input-wrapper,
                .${CSS_PREFIX}-custom-colors.${CSS_PREFIX}-select--theme-light .${CSS_PREFIX}-select__input-wrapper:focus
            `,
            rule: `box-shadow: inset 0 0 0 2px ${colorPalette.primary.L2}`,
        },
    ];
}

/**
 * Enhance isEmpty method to also works with numbers.
 *
 * @param  {any}     value The value to check.
 * @return {boolean} Weither if the input value is empty or != 0.
 */
const _isEmpty = (value) => {
    if (typeof value === 'number') {
        return value === 0;
    }

    return isEmpty(value);
};

/////////////////////////////
//                         //
//     Public functions    //
//                         //
/////////////////////////////

/**
 * Get the basic CSS class for the given type.
 *
 * @param  {string}         prefix The class name prefix for the generated CSS class.
 * @param  {string}         type   The type of CSS class we want to generate (e.g.: 'color', 'variant', ...).
 * @param  {string|boolean} value  The value of the type of the CSS class (e.g.: 'primary', 'button', ...).
 * @return {string}         The basic CSS class.
 */
function getBasicClass({ prefix, type, value }) {
    if (isBoolean(value)) {
        if (!value) {
            // False value should not return a class.
            return '';
        }
        const booleanPrefixes = ['has', 'is'];

        if (booleanPrefixes.some((booleanPrefix) => type.toString().startsWith(booleanPrefix))) {
            return `${prefix}--${kebabCase(type)}`;
        }

        return `${prefix}--is-${kebabCase(type)}`;
    }

    return `${prefix}--${kebabCase(type)}-${value}`;
}

/**
 * Return all basic LumX CSS classes which are available for every components.
 *
 * @see {@link /src/components/index.d.ts} for the possible values of each parameter.
 *
 * @param  {string} prefix The class name prefix for the generated CSS class.
 * @param  {Object} props  All the other props you want to generate a class.
 *                         The rule of thumb: the key is the name of the prop in the class, the value a string that will
 *                         be used in the classname to represent the value of the given prop.
 * @return {string} All LumX basic CSS classes.
 */
function handleBasicClasses({ prefix, ...props }) {
    const otherClasses = {};
    if (!isEmpty(props)) {
        Object.keys(props).forEach((prop) => {
            otherClasses[getBasicClass({ prefix, type: prop, value: props[prop] })] = isBoolean(props[prop])
                ? props[prop]
                : !_isEmpty(props[prop]);
        });
    }

    return classNames(prefix, otherClasses);
}

/**
 * Detects swipe direction.
 * Credits: http://javascriptkit.com/javatutors/touchevents2.shtml.
 *
 * @param  {Element}  el Element that will hold touch events.
 * @param  {Function} cb Callback function.
 * @return {Function} Function to remove listeners.
 */
function detectSwipe(el, cb = noop) {
    const touchsurface = el;
    let distX, distY, startX, startY, swipedir;
    // Required min distance traveled to be considered swipe.
    const threshold = 150;
    // Maximum distance allowed at the same time in perpendicular direction.
    const restraint = 100;
    // Maximum time allowed to travel that distance.
    const allowedTime = 300;
    let elapsedTime, startTime;
    const handleswipe = cb;

    const onTouchStart = (evt) => {
        const [touchobj] = evt.changedTouches;
        swipedir = 'none';
        // Const dist = 0;
        startX = touchobj.pageX;
        startY = touchobj.pageY;
        // Record time when finger first makes contact with surface.
        startTime = new Date().getTime();
        evt.preventDefault();
    };

    const onTouchMove = (evt) => {
        // Prevent scrolling when inside DIV.
        evt.preventDefault();
    };

    const onTouchEnd = (evt) => {
        const [touchobj] = evt.changedTouches;
        // Get horizontal dist traveled by finger while in contact with surface.
        distX = touchobj.pageX - startX;
        // Get vertical dist traveled by finger while in contact with surface.
        distY = touchobj.pageY - startY;
        // Get time elapsed.
        elapsedTime = new Date().getTime() - startTime;
        if (elapsedTime <= allowedTime) {
            // First condition for awipe met.
            if (Math.abs(distX) >= threshold && Math.abs(distY) <= restraint) {
                // 2nd condition for horizontal swipe met.
                // If dist traveled is negative, it indicates left swipe.
                swipedir = distX < 0 ? 'left' : 'right';
            } else if (Math.abs(distY) >= threshold && Math.abs(distX) <= restraint) {
                // 2nd condition for vertical swipe met.
                // If dist traveled is negative, it indicates up swipe.
                swipedir = distY < 0 ? 'up' : 'down';
            }
        }
        handleswipe(swipedir);
        evt.preventDefault();
    };

    touchsurface.addEventListener('touchstart', onTouchStart, false);
    touchsurface.addEventListener('touchmove', onTouchMove, false);
    touchsurface.addEventListener('touchend', onTouchEnd, false);

    return () => {
        touchsurface.removeEventListener('touchstart', onTouchStart, false);
        touchsurface.removeEventListener('touchmove', onTouchMove, false);
        touchsurface.removeEventListener('touchend', onTouchEnd, false);
    };
}

/**
 * Make sure the pressed key is the enter key before calling the callback.
 *
 * @param  {Function} cb The callback to call on enter/return press.
 * @return {Function} The decorated function.
 */
function onEnterPressed(cb) {
    return (evt) => {
        if (evt.key !== 'Enter') {
            return;
        }

        cb();
    };
}

/**
 * Make sure the pressed key is the escape key before calling the callback.
 *
 * @param  {Function} cb The callback to call on escape press.
 * @return {Function} The decorated function.
 */
function onEscapePressed(cb) {
    return (evt) => {
        if (evt.keyCode !== ESCAPE_KEY_CODE) {
            return;
        }

        cb();
    };
}

/**
 * Set custom color palette on primary and secondary colors.
 *
 * @param {Element} sheet        The sheet to insert the custom rules in.
 * @param {string}  theme        The theme to apply the custom color palete on.
 * @param {Object}  colorPalette The custom color palette.
 */
function setColorPalette(sheet, theme, colorPalette) {
    let index = 0;

    let buttonRules = [];
    buttonRules = buttonRules.concat(
        _getButtonCSSRules(colorPalette, 'primary'),
        _getButtonCSSRules(colorPalette, 'secondary'),
        _getButtonSelectedCSSRules(colorPalette),
    );

    buttonRules.forEach((buttonRule) => {
        _addCSSRule(sheet, buttonRule.selector, buttonRule.rule, index);
        index++;
    });

    let checkboxRules;
    if (theme === 'lumapps') {
        checkboxRules = _getCheckboxCSSRules(colorPalette, 'primary');
    } else if (theme === 'material') {
        checkboxRules = _getCheckboxCSSRules(colorPalette, 'secondary');
    }

    checkboxRules.forEach((checkboxRule) => {
        _addCSSRule(sheet, checkboxRule.selector, checkboxRule.rule, index);
        index++;
    });

    const chipRules = _getChipSelectedCSSRules(colorPalette);

    chipRules.forEach((chipRule) => {
        _addCSSRule(sheet, chipRule.selector, chipRule.rule, index);
        index++;
    });

    const progressRules = _getProgressCSSRules(colorPalette);

    progressRules.forEach((progressRule) => {
        _addCSSRule(sheet, progressRule.selector, progressRule.rule, index);
        index++;
    });

    let radioButtonRules;
    if (theme === 'lumapps') {
        radioButtonRules = _getRadioButtonCSSRules(colorPalette, 'primary');
    } else if (theme === 'material') {
        radioButtonRules = _getRadioButtonCSSRules(colorPalette, 'secondary');
    }

    radioButtonRules.forEach((radioButtonRule) => {
        _addCSSRule(sheet, radioButtonRule.selector, radioButtonRule.rule, index);
        index++;
    });

    const selectRules = _getSelectCSSRules(colorPalette);

    selectRules.forEach((selectRule) => {
        _addCSSRule(sheet, selectRule.selector, selectRule.rule, index);
        index++;
    });
}

/////////////////////////////

export { getBasicClass, handleBasicClasses, detectSwipe, onEnterPressed, onEscapePressed, setColorPalette };
