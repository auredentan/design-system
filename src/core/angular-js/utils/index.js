import './lx-depth_service';
import './lx-enter-keypress';
import './lx-event-scheduler_service';
import './lx-utils_service';
