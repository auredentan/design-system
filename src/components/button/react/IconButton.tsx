import React, { Children, cloneElement } from 'react';

import classNames from 'classnames';

import get from 'lodash/get';
import isEmpty from 'lodash/isEmpty';

import { Icon } from 'LumX';
import { COMPONENT_PREFIX } from 'LumX/core/react/constants';
import { ValidateParameters, getRootClassName, validateComponent } from 'LumX/core/react/utils';

import {
    Button,
    ButtonProps,
    CLASSNAME as BUTTON_CLASSNAME,
    Color,
    Colors,
    Emphasis,
    Emphasises,
    Size,
    Sizes,
    Theme,
    Themes,
} from './Button';

/////////////////////////////

enum Variants {
    icon = 'icon',
}
type Variant = Variants;

/////////////////////////////

/**
 * Defines the props of the component.
 */
interface IProps extends ButtonProps {
    /**
     * The <IconButton> should never have the `variant` prop as this prop is forced to 'icon' in the <Button>.
     */
    variant?: never;
}
type IconButtonProps = IProps;

/////////////////////////////

/**
 * Define the types of the default props.
 */
interface IDefaultPropsType extends Partial<IconButtonProps> {}

/////////////////////////////
//                         //
//    Public attributes    //
//                         //
/////////////////////////////

/**
 * The display name of the component.
 *
 * @type {string}
 * @constant
 * @readonly
 */
const COMPONENT_NAME: string = `${COMPONENT_PREFIX}IconButton`;

/**
 * The default class name and classes prefix for this component.
 *
 * @type {string}
 * @constant
 * @readonly
 */
const CLASSNAME: string = getRootClassName(COMPONENT_NAME);

/**
 * The default value of props.
 *
 * @type {IDefaultPropsType}
 * @constant
 * @readonly
 */
const DEFAULT_PROPS: IDefaultPropsType = {};

/////////////////////////////
//                         //
//    Private functions    //
//                         //
/////////////////////////////

/**
 * Globally validate the component before validating the children.
 *
 * @param {ValidateParameters} params The children, their number and the props of the component.
 */
function _preValidate({ props }: ValidateParameters): void {
    if (isEmpty(props.variant)) {
        return;
    }

    console.warn(
        `You shouldn't pass the \`variant\` prop in a <${COMPONENT_NAME}> as it's forced to 'icon' (got '${
            props.variant
        }')!`,
    );
}

/**
 * Validate the component props and children.
 * Also, sanitize, cleanup and format the children and return the processed ones.
 *
 * @param  {IconButtonProps} props The children and props of the component.
 * @return {React.ReactNode} The processed children of the component.
 */
function _validate(props: IconButtonProps): React.ReactNode {
    return validateComponent(COMPONENT_NAME, {
        allowedTypes: [Icon],
        maxChildren: 1,
        minChildren: 1,
        preValidate: _preValidate,
        props,
    });
}

/////////////////////////////

/**
 * Displays an icon button.
 * It's like a <Button> but only displays an icon instead of a label in the body of the button.
 *
 * Note that you cannot use the `variant` prop in this component.
 *
 * @see {@link Button} for more information on <Button>.
 *
 * @return {React.ReactElement} The component.
 */
const IconButton: React.FC<IconButtonProps> = ({
    children,
    className = '',
    ...props
}: IconButtonProps): React.ReactElement => {
    const newChildren: React.ReactNode = _validate({ children, ...props });

    return (
        <Button className={classNames(className, CLASSNAME)} {...props} variant={Variants.icon}>
            {Children.map(
                newChildren,
                // tslint:disable-next-line: no-any
                (child: any): React.ReactNode => {
                    return cloneElement(child, {
                        className: classNames(get(child.props, 'className', ''), `${BUTTON_CLASSNAME}__icon`),
                    });
                },
            )}
        </Button>
    );
};
IconButton.displayName = COMPONENT_NAME;

/////////////////////////////

export {
    CLASSNAME,
    DEFAULT_PROPS,
    Color,
    Colors,
    Emphasis,
    Emphasises,
    IconButton,
    IconButtonProps,
    Size,
    Sizes,
    Theme,
    Themes,
    Variant,
    Variants,
};
