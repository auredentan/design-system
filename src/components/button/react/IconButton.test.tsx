import React from 'react';

import { mount, shallow } from 'enzyme';
import mockConsole from 'jest-mock-console';
import { build, fake, oneOf } from 'test-data-bot';

import { Button, ButtonVariants } from 'LumX';
import { ICommonSetup, Wrapper, commonTestsSuite } from 'LumX/core/testing/utils.test';
import { mdiChevronDown, mdiPlus } from 'LumX/icons';

import { CLASSNAME, Emphasises, IconButton, IconButtonProps, Sizes, Themes } from './IconButton';

/////////////////////////////

/**
 * Define the overriding properties waited by the `setup` function.
 */
type ISetupProps = Partial<IconButtonProps>;

/**
 * Defines what is returned by the setup function.
 */
interface ISetup extends ICommonSetup {
    /**
     * The properties of the tested component.
     */
    props: ISetupProps;

    /**
     * The <Button> element that is used as a wrapper for the children of the <IconButton>.
     */
    button: Wrapper;
}

/////////////////////////////

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param props The props to use to override the default props of the component.
 * @param     [shallowRendering=true] Indicates if we want to do a shallow or a full rendering.
 * @return      An object with the props, the component wrapper and some shortcut to some element inside of
 *                       the component.
 */
const setup: (props?: ISetupProps, shallowRendering?: boolean) => ISetup = (
    { ...propsOverrides }: ISetupProps = {},
    shallowRendering: boolean = true,
): ISetup => {
    const props: IconButtonProps = {
        icon: mdiPlus,
        ...propsOverrides,
    };

    const renderer: (el: React.ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    const wrapper: Wrapper = renderer(<IconButton {...props} />);

    return {
        button: wrapper.find('Button'),

        props,
        wrapper,
    };
};

describe(`<${IconButton.displayName}>`, (): void => {
    // 1. Test render via snapshot (default state of component).
    describe('Snapshots and structure', (): void => {
        it('should render correctly an icon button', (): void => {
            const { button, wrapper }: ISetup = setup();
            expect(wrapper).toMatchSnapshot();

            expect(button).toExist();
            expect(button).toHaveClassName(CLASSNAME);

            expect(button).toHaveProp('leftIcon');
        });
    });

    /////////////////////////////

    // 2. Test defaultProps value and important props custom values.
    describe('Props', (): void => {
        it('should use default props', (): void => {
            const { button, props }: ISetup = setup();

            expect(button).toHaveProp('leftIcon', props.icon);
            expect(button).toHaveProp('variant', ButtonVariants.icon);
        });

        it("should use 'icon' `variant` whatever the given `variant` prop is", (): void => {
            mockConsole();

            const modifiedProps: ISetupProps = {
                // We known that <IconButton> could not have a `variant` prop.
                // @ts-ignore
                variant: ButtonVariants.icon,
            };

            let { button }: ISetup = setup(modifiedProps);

            expect(button).toHaveProp('variant', ButtonVariants.icon);

            /////////////////////////////

            // We known that <IconButton> could not have a `variant` prop.
            // @ts-ignore
            modifiedProps.variant = ButtonVariants.button;

            ({ button } = setup(modifiedProps));

            expect(button).toHaveProp('variant', ButtonVariants.icon);
        });

        it(`should forward any <${Button.displayName}> prop (except \`variant\`)`, (): void => {
            const modifiedPropsBuilder: () => ISetupProps = build('props').fields({
                // tslint:disable-next-line: no-any
                color: fake((fakeData: any): string => fakeData.commerce.color()),
                emphasis: oneOf(...Object.values(Emphasises)),
                icon: oneOf(mdiPlus, mdiChevronDown),
                size: oneOf(...Object.values(Sizes)),
                theme: oneOf(...Object.values(Themes)),
            });

            const modifiedProps: ISetupProps = modifiedPropsBuilder();

            const { button }: ISetup = setup(modifiedProps);

            Object.keys(modifiedProps).forEach(
                (prop: string): void => {
                    if (prop === 'icon') {
                        return;
                    }

                    expect(button).toHaveProp(prop, modifiedProps[prop]);
                },
            );
        });
    });

    /////////////////////////////

    // 3. Test events.
    describe('Events', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', (): void => {
        beforeEach(
            (): void => {
                try {
                    // If `console.warn` has been mocked at least one, this exists. So disable TS here.
                    // @ts-ignore
                    global.console.warn.mockRestore();
                } catch (exception) {
                    // Nothing to do here.
                }
            },
        );

        it('should fail when any child is given', (): void => {
            expect(
                (): void => {
                    setup({ children: 'Label', icon: mdiPlus });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it('should fail when no `icon` are given', (): void => {
            expect(
                (): void => {
                    setup({ icon: undefined });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it('should fail when using `leftIcon` instead of `icon`', (): void => {
            expect(
                (): void => {
                    // @ts-ignore
                    setup({ leftIcon: mdiPlus });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it('should fail when using `rightIcon` instead of `icon`', (): void => {
            expect(
                (): void => {
                    // @ts-ignore
                    setup({ rightIcon: mdiPlus });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it(`should warn the user when rendering a <${IconButton.displayName}> with a \`variant\``, (): void => {
            global.console.warn = jest.fn();

            // We know that a <IconButton> cannot receive a `variant`, but for the purpose of the test ignore it.
            // @ts-ignore
            setup({ variant: ButtonVariants.icon });
            expect(global.console.warn).toHaveBeenCalled();

            // @ts-ignore
            global.console.warn.mockClear();

            // We know that a <IconButton> cannot receive a `variant`, but for the purpose of the test ignore it.
            // @ts-ignore
            setup({ variant: ButtonVariants.button });
            expect(global.console.warn).toHaveBeenCalled();
        });
    });

    /////////////////////////////

    // 5. Test state.
    describe('State', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // Common tests suite.
    commonTestsSuite(setup, { className: 'button', prop: 'button' }, { className: CLASSNAME });
});
