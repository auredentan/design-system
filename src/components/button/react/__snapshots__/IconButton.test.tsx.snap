// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton> Conditions should fail when anything else than <Icon> is given as child 1`] = `"You can only have <Icon> children in <IconButton> (got <div>)!"`;

exports[`<IconButton> Conditions should fail when more than 1 children are given 1`] = `"<IconButton> must have exactly 1 child (got 2)!"`;

exports[`<IconButton> Conditions should fail when no child is given 1`] = `"<IconButton> must have exactly 1 child (got 0)!"`;

exports[`<IconButton> Snapshots and structure should render correctly an icon button 1`] = `
<Button
  className="lumx-icon-button"
  variant="icon"
>
  <Icon
    className="lumx-button__icon"
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
</Button>
`;
